src/directives/run-async.ts

```diff
@@ -0,0 +1,153 @@
+                                  }) => (part: NodePart) => {
+  const {success, pending, initial, failure} = templates;
+  const currentRunState = runs.get(part);
+
+  // The first time we see a value we save and await the work function.
+  // TODO(justinfagnani): allow a custom invalidate function
+  if (currentRunState === undefined || currentRunState.key !== key) {
+    // Abort a pending request if there is one
+    if (currentRunState !== undefined && currentRunState.state === 'pending') {
+      if (currentRunState.abortController !== undefined) {
+        currentRunState.abortController.abort();
+      }
+      // TODO(justinfagnani): This should be an AbortError, but it's not
+      // implemented yet
+      currentRunState.rejectPending(new Error());
+    }
+    const abortController =
+        hasAbortController ? new AbortController() : undefined;
+    const abortSignal =
+        hasAbortController ? abortController!.signal : undefined;
+    let resolvePending!: () => void;
+    let rejectPending!: (e: Error) => void;
+    const pendingPromise = new Promise((res, rej) => {
+      resolvePending = res;
+      rejectPending = rej;
+    });
+    const promise = task(key, {signal: abortSignal});
+    // The state is immediately 'pending', since the function has been
+    // executed, but if the function throws an InitialStateError to
+    // indicate that it couldn't even start processing, then we will set
+    // the state to 'initial'.
+    const runState: AsyncRunState = {
+      key,
+      promise,
+      state: 'pending',
+      abortController,
+      resolvePending,
+      rejectPending,
+    };
+    runs.set(part, runState);
+
+    Promise.resolve(promise).then(
+        (value: unknown) => {
+          runState.state = 'success';
+          const currentRunState = runs.get(part);
+          runState.resolvePending();
+          if (currentRunState !== runState) {
+            return;
+          }
+          part.setValue(success(value));
+          part.commit();
+        },
+        (error: Error) => {
+          const currentRunState = runs.get(part);
+          runState.rejectPending(new Error());
+          if (currentRunState !== runState) {
+            return;
+          }
+          if (error instanceof InitialStateError &&
+              typeof initial === 'function') {
+            runState!.state = 'initial';
+            part.setValue(initial());
+            part.commit();
+          } else {
+            runState!.state = 'failure';
+            if (typeof failure === 'function') {
+              // render success callback
+              part.setValue(failure(error));
+              part.commit();
+            }
+          }
+        });
+
+    (async () => {
+      // Wait a microtask for the initial render of the Part to complete
+      await 0;
+      const currentRunState = runs.get(part);
+      if (currentRunState === runState && currentRunState.state === 'pending') {
```

Comment:
If you only dispatch a promise when `currentRunState.state === 'pending'` an action that immediate throws `new InitialStateError()` won't be provided the `pendingPromise` reference to manage the reject that happens on [line 108](https://github.com/Polymer/lit-html/pull/657/files#diff-47605b25f29d88148b59b1949601eb08R108). Should this be updated to:
```suggestion
      if (currentRunState === runState && (currentRunState.state === 'pending' || currentRunState.state === 'initial')) {
```